### ISSUE TITLE
Prefetching Values: Add code in prehandle to take advantage of warm call in virtual map

### ISSUE THREAD

**Issue by nickpoorman** [hashgraph/hedera-services#3968]
Deepak has a PR ready that he hasn’t merged yet. He is doing some performance tests first to confirm that it can be merged with a positive performance impact. Once that’s merged, there may be some changes needed on services side to take advantage of the warm call.

Here's the related PR https://github.com/swirlds/swirlds-platform/pull/5902/files#diff-369a0d8347c5201620a3c7646eda2927953431a936cdab6387ba528f6b5f6186

**Comment by jasperpotts** [comment]
The issue is when we are processing transactions in handleTransaction that all happens sequentially on a single thread. As we move to more state on disk there are many disk reads that happen in that single thread. Each has a latency taking time and all that time adds up and limits the number of transactions we can handle per second.

**Comment by mhess-swl** [comment]
@rbair23 and @jasperpotts 

* Should we implement caching across all transactions types, or are we focusing on NFTs only?
* Under what conditions should we evict from the cache? Maybe an LRU cache that evicts every N transactions? 
	- Since this is POC-esque I would think we'd not bother building our own cache right now (to make the code available to @OlegMazurov sooner), but I thought I should ask: do we implement our own simple cache?
* What data are we caching? I'm assuming we're doing this for the NFT storage only, but if not--accounts, signatures, etc...? What will function as the key?

**Comment by OlegMazurov** [comment]
@mhess-swl You are not to implement a new caching solution but take advantage of a recently added API: `VirtualMap.warm(K key)`.
All data structures currently using `VirtualMap` are subject to warming up from the new background thread pool during pre-handle (though not all `VirtualMaps` are currently enabled by default). Keys are already defined.

**Comment by mhess-swl** [comment]
Another question related to [this comment](https://github.com/swirlds/swirlds-platform/blob/f9a9152cbe06b1300013f2eae5b216b5f1c6c05e/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/VirtualMap.java#L492-L494) from `VirtualMap.warm`: how do we know `what leaf records and internal records are going to be accessed`?

**Comment by lbaird** [comment]
The transaction contains accounts and contracts that will be touched. It’s
possible a contract will end up touching even more. But we can at least
warm the ones we can see.

**Comment by mhess-swl** [comment]
After discussing today, we can't add `warm(...)` calls on the thread that calls prehandle because there will be significant performance issues. Instead we've decided to use a dedicated threadpool that will call `warm(...)` on the appropriate entities, and prehandle will only enqueue events/transactions to that dedicated threadpool. The threadpool will call `warm`, and the transaction handler will then access the (hopefully pre-warmed) virtual map. @rbair23 @OlegMazurov @jasperpotts please verify I've understood :)

**Comment by nickpoorman** [comment]
Update: Oleg is testing and adding commits. Currently ongoing.

**Comment by mhess-swl** [comment]
I haven't been tracking the work we've been doing here, but I've had a number of conversations with @OlegMazurov to make improvements over the last few weeks. See ticket #4824 for a ticket that has the separate commits submitted. 

At the moment Oleg has discovered an issue with his benchmark not being randomized, which he's going to fix and then re-run.

**Comment by mhess-swl** [comment]
Created issue #5052 for the next changes needed. We are warming the cache too early; we need to move the cache warming from event intake to processing in parallel with the consensus thread (see the issue for more details).

**Comment by nickpoorman** [comment]
Let's do this in the old mono service code for now.

**Comment by mhess-swl** [comment]
This has now been merged to `develop` with the completion of #5868. @OlegMazurov is there anything else to do to keep this ticket open? Do you want to continue to track improvements here, or open a new ticket (or tickets) for those improvements?

**Comment by mhess-swl** [comment]
After a quick conversation with @OlegMazurov, we decided he'll open further tickets to tweak the cache warming implementation.
